File: android/core/src/main/kotlin/com/tickernest/core/Format.kt
package com.tickernest.core

import java.math.BigDecimal
import java.math.RoundingMode

/**
 * Money formatting. Same rules as the web client (Indian lakh/crore for INR,
 * comma-thousands for USD). Tabular numerals + 2 fractional digits always.
 *
 * Implementation note: we DO NOT route through Double/Long, because INR
 * portfolios past 2^53 lose precision. We chunk the integer string ourselves.
 */
fun formatMoney(v: Money, currency: Currency = Currency.INR): String {
    val abs = v.abs().setScale(2, RoundingMode.HALF_EVEN).toPlainString()
    val (intPart, fracPart) = abs.split('.').let { it[0] to (it.getOrNull(1) ?: "00") }
    val grouped = when (currency) {
        Currency.INR -> indianGroup(intPart)
        Currency.USD -> westernGroup(intPart)
    }
    val sign = if (v.signum() < 0) "-" else ""
    return "$sign$grouped.$fracPart"
}

fun formatPct(v: Money, places: Int = 2): String {
    val pct = v.multiply(BigDecimal(100)).setScale(places, RoundingMode.HALF_EVEN)
    val sign = if (pct.signum() > 0) "+" else ""
    return "$sign${pct.toPlainString()}%"
}

fun formatSignedMoney(v: Money, currency: Currency = Currency.INR): String {
    val formatted = formatMoney(v, currency)
    return if (v.signum() > 0) "+$formatted" else formatted
}

fun formatQty(v: Money): String {
    val s = v.setScale(4, RoundingMode.HALF_EVEN).toPlainString()
    return s.trimEnd('0').trimEnd('.').ifEmpty { "0" }
}

enum class Currency { INR, USD }

private fun indianGroup(intStr: String): String {
    if (intStr.length <= 3) return intStr
    val tail = intStr.substring(intStr.length - 3)
    var s = intStr.substring(0, intStr.length - 3)
    val parts = mutableListOf<String>()
    while (s.length > 2) {
        parts.add(0, s.substring(s.length - 2))
        s = s.substring(0, s.length - 2)
    }
    if (s.isNotEmpty()) parts.add(0, s)
    return "${parts.joinToString(",")},$tail"
}

private fun westernGroup(intStr: String): String {
    if (intStr.length <= 3) return intStr
    val parts = mutableListOf<String>()
    var s = intStr
    while (s.length > 3) {
        parts.add(0, s.substring(s.length - 3))
        s = s.substring(0, s.length - 3)
    }
    if (s.isNotEmpty()) parts.add(0, s)
    return parts.joinToString(",")
}

File: web/src/lib/format.ts
/**
 * Money formatting. Inputs are always strings (NUMERIC). We use Decimal-aware
 * grouping; never rely on Number() which loses precision at INR 100Cr+.
 */
import Decimal from 'decimal.js';

const inrIntFmt = new Intl.NumberFormat('en-IN', { maximumFractionDigits: 0 });
const usdIntFmt = new Intl.NumberFormat('en-US', { maximumFractionDigits: 0 });

export const formatMoney = (
  v: string | Decimal,
  currency: 'INR' | 'USD' = 'INR',
): string => {
  const d = new Decimal(v);
  const fmt = currency === 'INR' ? inrIntFmt : usdIntFmt;
  const s = d.abs().toFixed(2);
  const [int, frac = '00'] = s.split('.') as [string, string];
  // Number can lose precision past 2^53; format the integer string by chunking
  // ourselves for INR (Indian grouping) when it exceeds 15 digits.
  const grouped =
    int.length > 15
      ? indianGroup(int)
      : fmt.format(Number(int));
  const sign = d.isNegative() ? '-' : '';
  return `${sign}${grouped}.${frac}`;
};

const indianGroup = (intStr: string): string => {
  // Indian grouping: last 3, then commas every 2.
  if (intStr.length <= 3) return intStr;
  const head = intStr.slice(0, intStr.length - 3);
  const tail = intStr.slice(intStr.length - 3);
  const parts: string[] = [];
  let s = head;
  while (s.length > 2) {
    parts.unshift(s.slice(s.length - 2));
    s = s.slice(0, s.length - 2);
  }
  if (s) parts.unshift(s);
  return `${parts.join(',')},${tail}`;
};

export const formatPct = (v: string | Decimal, places = 2): string => {
  const d = new Decimal(v).mul(100);
  const sign = d.isPositive() && !d.isZero() ? '+' : '';
  return `${sign}${d.toFixed(places)}%`;
};

export const formatSignedMoney = (
  v: string | Decimal,
  currency: 'INR' | 'USD' = 'INR',
): string => {
  const d = new Decimal(v);
  const formatted = formatMoney(d, currency); // already prefixes '-' for negatives
  if (d.isPositive() && !d.isZero()) return `+${formatted}`;
  return formatted;
};

export const trendClass = (v: string | Decimal): string => {
  const d = new Decimal(v);
  if (d.isZero()) return 'text-flat';
  return d.isPositive() ? 'text-gain' : 'text-loss';
};

export const formatQty = (v: string | Decimal): string => {
  const d = new Decimal(v);
  // strip trailing zeros after decimal, max 4 places
  return d.toDecimalPlaces(4).toString();
};
